singularity-eos/eos/eos_base.hpp

```diff
@@ -657,13 +670,76 @@ class EosBase {
     PORTABLE_ALWAYS_THROW_OR_ABORT(msg);
   }
 
+  // Serialization
+  // JMM: These must frequently be special-cased.
+  std::size_t DynamicMemorySizeInBytes() const { return 0; }
+  std::size_t DumpDynamicMemory(char *dst) { return 0; }
+  std::size_t SetDynamicMemory(char *src,
+                               const SharedMemSettings &stngs = DEFAULT_SHMEM_STNGS) {
+    return 0;
+  }
+  // JMM: These usually don't need to be special cased.
+  std::size_t SharedMemorySizeInBytes() const {
+    const CRTP *pcrtp = static_cast<const CRTP *>(this);
+    return pcrtp->DynamicMemorySizeInBytes();
+  }
+  constexpr bool AllDynamicMemoryIsShareable() const { return true; }
+  // JMM: These are generic and never need to be special-cased.
+  std::size_t SerializedSizeInBytes() const {
+    // sizeof(*this) returns the size of JUST the base class.
+    const CRTP *pcrtp = static_cast<const CRTP *>(this);
+    std::size_t dyn_size = pcrtp->DynamicMemorySizeInBytes();
+    return dyn_size + sizeof(CRTP);
```

Comment:
hmmm... won't this give the size of the underlying class rather than the size of the variant? Without thinking too hard, I'd think we want the size of the variant, maybe that's the subtlety in the 2 size functions ?

Reply:
I'm still confused, shouldn't `SharedMemorySizeInBytes()` just return `sizeof(*this)` in the variant implementation rather than calling the held type's version?

Reply:
There might be some confusion as to the model here. When I call `Serialize` from either the variant or a bare class, I should get a `char` pointer pointing to serialized memory and a size. That size should be the information required to rebuild the tables and set the object, whether it be a variant or a bare class instance. The main difference here is that the variant `sizeof(*this)` and the bare class instance `sizeof(*this)` are different because the variant has a bigger memory footprint and different layout. So you need separate methods per class for dealing with the tables and then the variant and the base class have different serialization/deserialization routines for the trivially-copyable memory that account for the fact that the type is different.

`SharedMemory` is different. This is the amount of *non*-trivially copyable memory that can be written to/read from the shared allocation between MPI ranks. It's the part that's not thread local. *objects* are thread local so their metadata must be too. The only thing that goes in shared memory is tables. `SharedMemorySizeInBytes` is the same as `DynamicMemorySizeInBytes` for every object except EOSPAC. This is mainly because EOSPAC is a library. So it's serialized contents include not only tables but also metadata that are not contained in `sizeof(EOSPAC)`.

Reply:
Discussed with @dholladay00 offline. Resolving.